src/Picasso_ParticleInit.hpp

```diff
@@ -519,6 +519,145 @@ void initializeParticlesSurface( InitRandom, const ExecutionSpace&,
+                                 const InitFunc& create_functor,
+                                 ParticleListType& surface_particle_list )
+{
+    Kokkos::Profiling::pushRegion( "Picasso::initializeParticles::Surface" );
+
+    // Particle type.
+    using particle_type = typename ParticleListType::particle_type;
+
+    // Get the local grid.
+    const auto& local_grid = *( surface_particle_list.mesh().localGrid() );
+
+    // Create a local mesh.
+    auto local_mesh = Cajita::createLocalMesh<ExecutionSpace>( local_grid );
+
+    // Get the global grid.
+    const auto& global_grid = local_grid.globalGrid();
+
+    // Get the particles.
+    auto& particles = surface_particle_list.aosoa();
+
+    // Allocate the particles.
+    int num_facets = surface.facets.extent( 0 );
+    int num_particles = particles_per_facet * num_facets;
+    particles.resize( num_particles );
+
+    // FIXME: Re-thread over particles instead of facets.
+    // Initialize particles.
+    Kokkos::parallel_for(
+        "Picasso::ParticleInit::RandomSurface",
+        Kokkos::RangePolicy<ExecutionSpace>( 0, num_facets ),
+        KOKKOS_LAMBDA( const int f ) {
+            // Particle coordinate.
+            Vec3<double> px;
+
+            // Particle.
+            particle_type particle;
+
+            Vec3<double> a = { surface.facets( f, 0, 0 ),
+                               surface.facets( f, 0, 1 ),
+                               surface.facets( f, 0, 2 ) };
+
+            Vec3<double> b = { surface.facets( f, 1, 0 ),
+                               surface.facets( f, 1, 1 ),
+                               surface.facets( f, 1, 2 ) };
+
+            Vec3<double> c = { surface.facets( f, 2, 0 ),
+                               surface.facets( f, 2, 1 ),
+                               surface.facets( f, 2, 2 ) };
+
+            Vec3<double> ab = b - a;
+            Vec3<double> ac = c - a;
+
+            auto cross = ab % ac;
+            double c2 = ~cross * cross;
+            double sqc2 = sqrt( c2 );
+            double facet_area = 0.5 * sqc2;
+
+            Vec3<double> pan =
+                cross / sqc2; // Create unit normal vector for the facet area
+
+            // Particle surface area.
+            double pa = facet_area / particles_per_facet;
+
+            // centroid of triangle facet
+            Vec3<double> centroid = ( a + b + c ) / 3.0;
+	    Vec3<double> a_centroid = ( a + centroid ) / 2.0;
+	    Vec3<double> b_centroid = ( b + centroid ) / 2.0;
+	    Vec3<double> c_centroid = ( c + centroid ) / 2.0;
+	    Kokkos::Array<Vec3<double>, 4> abc_centroid 
+	                             = { centroid, a_centroid, b_centroid, c_centroid };
+
+            for ( int ip = 0; ip < particles_per_facet; ++ip )
+            {
+                // Local particle id.
+                int pid = f * particles_per_facet + ip;
+
+                // ppf = 3
+		if( particles_per_facet == 3 )
+		   px = abc_centroid[ip+1];
+		// ppf = 1 or 4
+		else
```

Comment:
I updated here to get the style that matches the internal particle uniform init. For `ppf=0` you get one on the centroid, then the intent is to add 1 per facet median for each `ppf>0`, but those are wrong. 

@abisner can you take a look to fix since you wrote the random version? Mostly looking for anything reasonable we can merge for now, even if we only really use the one on the centroid for now

Reply:
@streeve I fixed it to work with internal division points with particles_per_facet_median if this is your intention

Reply:
No, your recursive version is better